src/Radicle/Internal/Core.hs

```diff
@@ -601,6 +589,26 @@ specialForms = Map.fromList $ first Ident <$>
           pure nil
         _ -> throwErrorHere $ OtherError "def-rec can only be used to define functions"
 
+-- Given a list of forms, the first three of which should be module metadata,
+-- runs the rest of the forms in a new scope, and then defs the module value.
+createModule :: Monad m => Text -> [Value] -> Lang m Value
+createModule ctx = \case
+    (name : doc : exports : forms) -> do
+      name' <- baseEval name
+      doc' <- baseEval doc
+      exports' <- baseEval exports
+      case (name', doc', exports') of
+        (Atom n, String d, Vec is) -> do
+          is' <- traverse isAtom is ?? toLangError (OtherError $ ctx <> ": Module export vector should only contain atoms")
+          e <- withEnv identity $ traverse_ eval forms *> gets bindingsEnv
```

Comment:
Is this right? Shouldn't it be `withEnv [im]pureEnv`? I'd think that modules are evaluated in a blank environment, so that if they "work" in one place they always work.

Reply:
(Sorry, I meant `withEnv (const pureEnv)`)

Reply:
- The current way a module uses code from another module is by `(import another)` where usually `another` is an identifier referring to a previously `def`ined module. So if the module was ran in a pure env, it wouldn't find this. If a module is to start from a totally pure env, I don't see how we can avoid running modules it depends on multiple times (if e.g. `a` and `b` both depend on `c`).

- Modules mostly assume the env has already been changed, e.g. `(:test ...)` assumes the `eval` has been changed.

- I quite like the dynamic nature. E.g. I can do:

```clojure
(def counter
  (fn [starts-at]
    (def i (ref starts-at))
    (def next ...)
    (def current ...)
    (module '[next current])))

...

(import (counter 42) issue-id)
(def my-issue-id (issue-id/next))
```